packages/central-server/app/admin/users.js

```diff
@@ -59,20 +195,429 @@ usersRouter.post(
+      await UserFacility.bulkCreate(
+        uniqueFacilityIds.map(facilityId => ({
+          userId: user.id,
+          facilityId,
+        })),
+        {
+          ignoreDuplicates: true,
+        },
+      );
+    });
+
+    res.send({ ok: true });
+  }),
+);
+
+const VALIDATION = yup
+  .object()
+  .shape({
+    displayName: yup.string().trim().required(),
+    email: yup.string().trim().email().required(),
+  })
+  .noUnknown();
+
+usersRouter.post(
+  '/validate',
+  asyncHandler(async (req, res) => {
+    const {
+      store: {
+        models: { User },
       },
     } = req;
 
     req.checkPermission('create', 'User');
 
     const fields = await VALIDATION.validate(req.body);
+
+    const existingUserWithSameEmail = await User.findOne({
+      where: {
+        email: fields.email,
+      },
+    });
+
+    const existingUserWithSameDisplayName = await User.findOne({
+      where: {
+        displayName: { [Op.iLike]: fields.displayName },
+      },
+    });
+
+    res.send({
+      isEmailUnique: !existingUserWithSameEmail,
+      isDisplayNameUnique: !existingUserWithSameDisplayName,
+    });
+  }),
+);
+
+const UPDATE_VALIDATION = yup
+  .object()
+  .shape({
+    visibilityStatus: yup
+      .string()
+      .required()
+      .oneOf([VISIBILITY_STATUSES.CURRENT, VISIBILITY_STATUSES.HISTORICAL]),
+    displayName: yup.string().trim().required(),
+    displayId: yup.string().trim().nullable().optional(),
+    role: yup.string().required(),
+    phoneNumber: yup.string().trim().nullable().optional(),
+    email: yup.string().trim().email().required(),
+    designations: yup.array().of(yup.string()).nullable().optional(),
+    newPassword: yup.string().nullable().optional(),
+    confirmPassword: yup.string().nullable().optional(),
+    allowedFacilityIds: yup.array().of(yup.string()).nullable().optional(),
+  })
+  .test('passwords-match', 'Passwords must match', function (value) {
+    const { newPassword, confirmPassword } = value;
+    // If both passwords are provided, they must match
+    if ((newPassword || confirmPassword) && newPassword !== confirmPassword) {
+      return this.createError({ message: 'Passwords must match' });
+    }
+    return true;
+  })
+  .test('password-is-not-hashed', 'Password must not be hashed', function (value) {
+    return !isBcryptHash(value.newPassword);
+  })
+  .noUnknown();
+usersRouter.put(
+  '/:id',
+  asyncHandler(async (req, res) => {
+    const {
+      store: {
+        models: { Role, User, UserDesignation, ReferenceData, UserFacility },
+      },
+      params: { id },
+      db,
+    } = req;
+
+    const fields = await UPDATE_VALIDATION.validate(req.body);
+
+    const user = await User.findByPk(id);
+    if (!user) {
+      throw new NotFoundError('User not found');
+    }
+    // only allow updating the user if the user has the write permission for the all users
+    req.checkPermission('write', subject('User', { id: String(Date.now()) }));
```

Comment:
Here we already have the user object so we can just pass that without using `subject`:


```suggestion
    req.checkPermission('write', user);
```